File: src/treemapper/diffctx/edges/semantic/erlang.py
from __future__ import annotations

import re
from collections import defaultdict
from pathlib import Path

from ...config.weights import EDGE_WEIGHTS
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict, discover_files_by_refs

_ERLANG_EXTS = frozenset({".erl", ".hrl"})

_MODULE_RE = re.compile(r"^-module\(([a-z_][a-zA-Z0-9_]{0,200})\)\.", re.MULTILINE)
_INCLUDE_RE = re.compile(r'^-include\("([^"]{1,300})"\)\.', re.MULTILINE)
_INCLUDE_LIB_RE = re.compile(r'^-include_lib\("([^"]{1,300})"\)\.', re.MULTILINE)
_BEHAVIOUR_RE = re.compile(r"^-behaviou?r\(([a-z_][a-zA-Z0-9_]{0,200})\)\.", re.MULTILINE)
_IMPORT_RE = re.compile(r"^-import\(([a-z_][a-zA-Z0-9_]{0,200})\s*,", re.MULTILINE)
_REMOTE_CALL_RE = re.compile(r"\b([a-z_][a-zA-Z0-9_]{0,100}):([a-z_][a-zA-Z0-9_]{0,100})\s*\(")
_FUNCTION_DEF_RE = re.compile(r"^([a-z_][a-zA-Z0-9_]{0,100})\s*\(", re.MULTILINE)

_ERLANG_STDLIB_MODULES = frozenset(
    {
        "io",
        "lists",
        "maps",
        "string",
        "binary",
        "file",
        "gen_server",
        "gen_statem",
        "gen_event",
        "supervisor",
        "application",
        "ets",
        "timer",
        "erlang",
        "proplists",
        "dict",
        "sets",
        "gb_trees",
        "gb_sets",
        "queue",
        "ordsets",
        "orddict",
        "math",
        "re",
        "unicode",
        "calendar",
        "filename",
        "filelib",
        "os",
        "code",
        "sys",
        "proc_lib",
        "error_logger",
        "logger",
        "inet",
        "gen_tcp",
        "gen_udp",
        "ssl",
        "crypto",
        "rand",
        "httpc",
        "httpd",
    }
)


def _is_erlang_file(path: Path) -> bool:
    return path.suffix.lower() in _ERLANG_EXTS


def _extract_module_name(content: str) -> str | None:
    m = _MODULE_RE.search(content)
    return m.group(1) if m else None


def _extract_includes(content: str) -> set[str]:
    refs: set[str] = set()
    for m in _INCLUDE_RE.finditer(content):
        refs.add(m.group(1))
    for m in _INCLUDE_LIB_RE.finditer(content):
        refs.add(m.group(1))
    return refs


def _extract_behaviours(content: str) -> set[str]:
    return {m.group(1) for m in _BEHAVIOUR_RE.finditer(content)}


def _extract_imports(content: str) -> set[str]:
    return {m.group(1) for m in _IMPORT_RE.finditer(content)}


def _extract_remote_calls(content: str) -> set[str]:
    return {m.group(1) for m in _REMOTE_CALL_RE.finditer(content) if m.group(1) not in _ERLANG_STDLIB_MODULES}


def _include_to_filename(ref: str) -> str:
    return ref.split("/")[-1].lower()


class ErlangEdgeBuilder(EdgeBuilder):
    weight = 0.65
    include_weight = EDGE_WEIGHTS["erlang_include"].forward
    behaviour_weight = EDGE_WEIGHTS["erlang_behaviour"].forward
    call_weight = EDGE_WEIGHTS["erlang_call"].forward
    reverse_weight_factor = EDGE_WEIGHTS["erlang_include"].reverse_factor

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
    ) -> list[Path]:
        erl_changed = [f for f in changed_files if _is_erlang_file(f)]
        if not erl_changed:
            return []

        refs: set[str] = set()
        for f in erl_changed:
            try:
                content = f.read_text(encoding="utf-8")
                for inc in _extract_includes(content):
                    refs.add(_include_to_filename(inc))
                for mod in _extract_remote_calls(content):
                    refs.add(f"{mod}.erl")
                for mod in _extract_imports(content):
                    refs.add(f"{mod}.erl")
                for beh in _extract_behaviours(content):
                    refs.add(f"{beh}.erl")
            except (OSError, UnicodeDecodeError):
                continue

        return discover_files_by_refs(refs, changed_files, all_candidate_files, repo_root)

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        erl_frags = [f for f in fragments if _is_erlang_file(f.path)]
        if not erl_frags:
            return {}

        edges: EdgeDict = {}
        mod_to_frags = self._build_module_index(erl_frags)
        name_to_frags = self._build_name_index(erl_frags)

        for ef in erl_frags:
            self._link_fragment(ef, mod_to_frags, name_to_frags, edges)

        return edges

    def _build_module_index(self, erl_frags: list[Fragment]) -> dict[str, list[FragmentId]]:
        mod_to_frags: dict[str, list[FragmentId]] = defaultdict(list)
        for f in erl_frags:
            mod_name = _extract_module_name(f.content)
            if mod_name:
                mod_to_frags[mod_name.lower()].append(f.id)
            stem = f.path.stem.lower()
            mod_to_frags[stem].append(f.id)
        return mod_to_frags

    def _build_name_index(self, erl_frags: list[Fragment]) -> dict[str, list[FragmentId]]:
        name_to_frags: dict[str, list[FragmentId]] = defaultdict(list)
        for f in erl_frags:
            name_to_frags[f.path.name.lower()].append(f.id)
        return name_to_frags

    def _link_fragment(
        self,
        ef: Fragment,
        mod_to_frags: dict[str, list[FragmentId]],
        name_to_frags: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        self._link_includes(ef, name_to_frags, edges)
        self._link_behaviours(ef, mod_to_frags, edges)
        self._link_imports(ef, mod_to_frags, edges)
        self._link_remote_calls(ef, mod_to_frags, edges)

    def _link_includes(
        self,
        ef: Fragment,
        name_to_frags: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for inc in _extract_includes(ef.content):
            filename = _include_to_filename(inc)
            self.add_edges_from_ids(
                ef.id,
                name_to_frags.get(filename, []),
                self.include_weight,
                edges,
            )

    def _link_behaviours(
        self,
        ef: Fragment,
        mod_to_frags: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for beh in _extract_behaviours(ef.content):
            self.add_edges_from_ids(
                ef.id,
                mod_to_frags.get(beh.lower(), []),
                self.behaviour_weight,
                edges,
            )

    def _link_imports(
        self,
        ef: Fragment,
        mod_to_frags: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for mod in _extract_imports(ef.content):
            self.add_edges_from_ids(
                ef.id,
                mod_to_frags.get(mod.lower(), []),
                self.include_weight,
                edges,
            )

    def _link_remote_calls(
        self,
        ef: Fragment,
        mod_to_frags: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for mod in _extract_remote_calls(ef.content):
            self.add_edges_from_ids(
                ef.id,
                mod_to_frags.get(mod.lower(), []),
                self.call_weight,
                edges,
            )

File: diffctx/src/edges/semantic/erlang.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::weights::EDGE_WEIGHTS;
use crate::types::Fragment;

use super::super::base::{self, EdgeBuilder, add_edge, add_edges_from_ids, discover_files_by_refs};
use super::super::EdgeDict;

fn is_erlang_file(path: &Path) -> bool {
    let ext = base::file_ext(path);
    ext == ".erl" || ext == ".hrl"
}

static MODULE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^-module\((\w+)\)").unwrap());
static INCLUDE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^-include(?:_lib)?\(\s*"([^"]+)""#).unwrap());
static BEHAVIOUR_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^-behaviou?r\((\w+)\)").unwrap());
static IMPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^-import\((\w+),").unwrap());
static FUNC_DEF_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^([a-z]\w*)\s*\(").unwrap());
static REMOTE_CALL_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\b([a-z]\w*):(\w+)\s*\(").unwrap());
static EXPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^-export\(\[([^\]]+)\]\)").unwrap());

fn extract_refs(content: &str) -> FxHashSet<String> {
    let mut refs = FxHashSet::default();
    refs.extend(INCLUDE_RE.captures_iter(content).map(|c| c[1].to_string()));
    refs.extend(BEHAVIOUR_RE.captures_iter(content).map(|c| c[1].to_string()));
    refs.extend(IMPORT_RE.captures_iter(content).map(|c| c[1].to_string()));
    refs.extend(REMOTE_CALL_RE.captures_iter(content).map(|c| c[1].to_string()));
    refs
}

fn extract_modules(content: &str) -> FxHashSet<String> {
    MODULE_RE.captures_iter(content).map(|c| c[1].to_string()).collect()
}

fn extract_func_defs(content: &str) -> FxHashSet<String> {
    FUNC_DEF_RE.captures_iter(content).map(|c| c[1].to_string()).collect()
}

pub struct ErlangEdgeBuilder;

impl EdgeBuilder for ErlangEdgeBuilder {
    fn build(&self, fragments: &[Fragment], repo_root: Option<&Path>) -> EdgeDict {
        let frags: Vec<&Fragment> = fragments.iter().filter(|f| is_erlang_file(Path::new(f.path()))).collect();
        if frags.is_empty() { return FxHashMap::default(); }

        let include_w = EDGE_WEIGHTS["erlang_include"].forward;
        let behaviour_w = EDGE_WEIGHTS["erlang_behaviour"].forward;
        let call_w = EDGE_WEIGHTS["erlang_call"].forward;
        let reverse_factor = EDGE_WEIGHTS["erlang_include"].reverse_factor;

        let idx = base::FragmentIndex::new(fragments, repo_root);
        let mut mod_to_frags: FxHashMap<String, Vec<_>> = FxHashMap::default();
        let mut fn_to_frags: FxHashMap<String, Vec<_>> = FxHashMap::default();
        for f in &frags {
            for m in extract_modules(&f.content) {
                mod_to_frags.entry(m.to_lowercase()).or_default().push(f.id.clone());
            }
            for name in extract_func_defs(&f.content) {
                fn_to_frags.entry(name.to_lowercase()).or_default().push(f.id.clone());
            }
        }

        let mut edges: EdgeDict = FxHashMap::default();

        for f in &frags {
            let self_fns = extract_func_defs(&f.content);
            for cap in INCLUDE_RE.captures_iter(&f.content) {
                base::link_by_name(&f.id, &cap[1], &idx, &mut edges, include_w, reverse_factor);
            }
            for cap in BEHAVIOUR_RE.captures_iter(&f.content) {
                if let Some(targets) = mod_to_frags.get(&cap[1].to_lowercase()) {
                    add_edges_from_ids(&mut edges, &f.id, targets, behaviour_w, reverse_factor);
                }
            }
            for cap in IMPORT_RE.captures_iter(&f.content) {
                if let Some(targets) = mod_to_frags.get(&cap[1].to_lowercase()) {
                    add_edges_from_ids(&mut edges, &f.id, targets, include_w, reverse_factor);
                }
            }
            for cap in REMOTE_CALL_RE.captures_iter(&f.content) {
                let module = &cap[1];
                if let Some(targets) = mod_to_frags.get(&module.to_lowercase()) {
                    for t in targets { if t != &f.id { add_edge(&mut edges, &f.id, t, call_w, reverse_factor); } }
                }
            }
            for id in &f.identifiers {
                if self_fns.contains(id) { continue; }
                if let Some(targets) = fn_to_frags.get(&id.to_lowercase()) {
                    for t in targets { if t != &f.id { add_edge(&mut edges, &f.id, t, call_w, reverse_factor); } }
                }
            }
        }
        edges
    }

    fn discover_related_files(
        &self, changed: &[PathBuf], candidates: &[PathBuf],
        repo_root: Option<&Path>, file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let erl_changed: Vec<&PathBuf> = changed.iter().filter(|f| is_erlang_file(f)).collect();
        if erl_changed.is_empty() { return vec![]; }
        let mut refs = FxHashSet::default();
        for f in &erl_changed {
            if let Some(content) = base::read_file_cached(f, file_cache) {
                refs.extend(extract_refs(&content));
            }
        }
        discover_files_by_refs(&refs, changed, candidates, repo_root)
    }
}
